sdks/python/apache_beam/typehints/typecheck.py

```diff
@@ -265,3 +268,89 @@ def visit_transform(self, applied_transform):
                 transform.get_type_hints(),
                 applied_transform.full_label),
             applied_transform.full_label)
+
+
+class PerformanceTypeCheckVisitor(pipeline.PipelineVisitor):
+
+  _in_combine = False
+  combine_classes = (
+      core.CombineFn,
+      core.CombinePerKey,
+      core.CombineValuesDoFn,
+      core.CombineValues,
+      core.CombineGlobally)
+
+  def enter_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = True
+
+  def leave_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = False
+
+  def visit_transform(self, applied_transform):
+    transform = applied_transform.transform
+    if isinstance(transform, core.ParDo) and not self._in_combine:
+      # Prefix label with 'ParDo' if necessary
+      full_label = applied_transform.full_label
+      if not full_label.startswith('ParDo'):
+        full_label = 'ParDo(%s)' % full_label
+
+      # Store output type hints in current transform
+      transform.fn._runtime_output_constraints = {}
+      output_type_hints = self.get_output_type_hints(transform)
+      if output_type_hints:
+        transform.fn._runtime_output_constraints[full_label] = (
+            output_type_hints)
+
+      # Store input type hints in producer transform
```

Comment:
This does not need to be guarded by if `isinstance(transform, core.ParDo)`, we can do it for all transforms. (Actually, the same could be said of the output type hints if we eliminated the direct access to `_runtime_output_constraints` as mentioned above.

Reply:
If this results in test failures due to more (incorrect) type hints being applied, feel free to drop a TODO here for future work.

Reply:
It fails for `Impulse`, `GroupByKey`, `Flatten` and a few other transforms. Sounds good, added a TODO for now.

Reply:
Nevermind, fixed it :)